### PR TITLE
Add --create option to `flutter emulators` command

### PR DESCRIPTION
**Pull request by DanTup**
This is still a bit rough and has no tests, but I wanted to nail down the logic before I start adding tests against it. This is based on the logic described in https://github.com/flutter/flutter/issues/13379#issuecomment-394686208. I don't entirely like how we have to do it (for ex. trying to create on without the sdk ID because we can't get a list of these, only devices/targets) but (on my Macbook at least) it seems to work.

LMK what you think. We may also wish to review the device we pick (currently we try `pixel` and `pixel_xl`) and whether we want to lock to a specific Android API version or just take the highest (which is what this currently does).

screen shot 2018-06-06 at 1 13 35 pm

(cc @devoncarew @mit-mit)

### CODE REVIEW

File: packages/flutter_tools/lib/src/android/android_sdk.dart [hunk 1]
```diff
@@ -104,6 +104,23 @@ String getAvdPath() {
   );
 }
 
+/// Locate 'avdmanager'. Prefer to use one from an Android SDK, if we can locate that.
+/// This should be used over accessing androidSdk.avdManagerPath directly because it
+/// will work for those users who have Android Tools installed but
+/// not the full SDK.
+String getAvdManagerPath([AndroidSdk existingSdk]) {
```
Comment (devoncarew) [inline packages/flutter_tools/lib/src/android/android_sdk.dart]: use a named param instead?

File: packages/flutter_tools/lib/src/android/android_sdk.dart [hunk 2]
```diff
@@ -104,6 +104,23 @@ String getAvdPath() {
...
+  if (existingSdk?.avdManagerPath != null)
+    return existingSdk.avdManagerPath;
+
+  final AndroidSdk sdk = AndroidSdk.locateAndroidSdk();
+
+  if (sdk?.latestVersion == null) {
```
Comment (devoncarew) [inline packages/flutter_tools/lib/src/android/android_sdk.dart]: I suspect this fallback logic parallels other android tools lookups in the source base (adb?). We may want to just fail if we can't locate an android sdk however (not do the path based lookup for `avdmanager`).

File: packages/flutter_tools/lib/src/commands/emulators.dart [hunk 3]
```diff
@@ -16,6 +16,8 @@ class EmulatorsCommand extends FlutterCommand {
   EmulatorsCommand() {
     argParser.addOption('launch',
         help: 'The full or partial ID of the emulator to launch.');
+    argParser.addOption('create',
```
Comment (devoncarew) [inline packages/flutter_tools/lib/src/commands/emulators.dart]: Should `--create` be a flag (instead of an option)? We could auto-name the resulting image (`pixel`, `pixel_xl-2`, ...), and have an optional flag to name a created image.

Comment (devoncarew) [reply packages/flutter_tools/lib/src/commands/emulators.dart]: `create` feels like a subcommand, even if the `package:args` api doesn't make it easy to have it as one.

File: packages/flutter_tools/lib/src/commands/emulators.dart [hunk 4]
```diff
@@ -65,6 +69,18 @@ class EmulatorsCommand extends FlutterCommand {
...
 
+  Future<Null> _createEmulator(String name) async {
+    final CreateEmulatorResult createResult =
+        await emulatorManager.createEmulator(name);
+
+    if (createResult.success) {
+      printStatus("Emulator '$name' created successfully.");
+    } else {
+      printStatus(createResult.error);
+      printStatus("Failed to create emulator '$name'.");
```
Comment (devoncarew) [inline packages/flutter_tools/lib/src/commands/emulators.dart]: We should print some text here about where they can find more info about creating emulator images (referencing flutter.io, or Android / Android Studio docs).

**Review by devoncarew — COMMENTED** [review]
Thanks for starting this! I added some comments on the CLI UI.

File: packages/flutter_tools/lib/src/android/android_sdk.dart (hunk 1 above)
Comment (DanTup) [reply packages/flutter_tools/lib/src/android/android_sdk.dart]: This matches other functions in this file. Should we change them to keep them consistent?

```dart
String getAdbPath([AndroidSdk existingSdk]);
String getEmulatorPath([AndroidSdk existingSdk]);
```

File: packages/flutter_tools/lib/src/android/android_sdk.dart (hunk 2 above)
Comment (DanTup) [reply packages/flutter_tools/lib/src/android/android_sdk.dart]: Currently there are three possibilities in here (same for the other few in this file):

1. If it exists on the SDK passed in (which looks to come from `context[AndroidSdk]`), we use that
2. If we can locate an SDK, that's used
3. Otherwise we try `os.which('emulator')?.path`

I'm not sure which out 1 and 2 is the best one (or whether we need 3). WDYT? If going with 1 then we can just delete this function; but I'm not sure whether that's a good bet. There are some other paths that do seem to use that and bail out if it's not set without any fallback, eg.:

```dart
  Future<bool> _checkForSupportedAdbVersion() async {
    if (androidSdk == null)
      return false;
```

Though interestingly, it then calls `getAdbPath` which does the fallback; but in this case it would never trigger because of the `return` above.

File: packages/flutter_tools/lib/src/commands/emulators.dart (hunk 3 above)
Comment (DanTup) [reply packages/flutter_tools/lib/src/commands/emulators.dart]: > Should `--create be a flag` (instead of an option)?  We could auto-name the resulting image

I was a bit unsure about this. If we pick the name for the user then it feels weird if we tell them "an emulator with that name already exists" so I was going to make them always provide it.

Of course, we could do the check ourselves first and only prompt if it already exists, but that might be better handled in the editor than here so it has better control over how to present it.

I was wondering if we should expose an `emulator.getDeviceProfiles`. That way when the user chooses "create emulator" in the editor, we can fetch the possible profiles (and filter it to pixels, nexuses, whatever) and then let them pick a device. I think being able to pick between a phone and a tablet could be helpful?

File: packages/flutter_tools/lib/src/commands/emulators.dart (hunk 4 above)
Comment (DanTup) [reply packages/flutter_tools/lib/src/commands/emulators.dart]: How's this? Seems a bit weird sending people to Android Studio or avdmanager docs but since the errors come from them I'm not sure if we can provide the info ourselves.

screen shot 2018-06-11 at 6 00 30 pm

File: packages/flutter_tools/lib/src/android/android_sdk.dart (hunk 1 above)
Comment (devoncarew) [reply packages/flutter_tools/lib/src/android/android_sdk.dart]: No, let's keep them consistent.

File: packages/flutter_tools/lib/src/android/android_sdk.dart (hunk 2 above)
Comment (devoncarew) [reply packages/flutter_tools/lib/src/android/android_sdk.dart]: Let's just delete the `os.which('emulator')?.path` option.

> Though interestingly, it then calls getAdbPath which does the fallback; but in this case it would never trigger because of the return above.

Hmm, sounds like some logic was orphaned by a refactor. Falling back to 'which `adb`' was useful since the user could then have a little functionality - like listing devices - even if they didn't have the full functionality that configuring an android sdk would give. If this regressed at some point, and people didn't notice, than it's likely that we don't need the fallback even for `adb`.

File: packages/flutter_tools/lib/src/commands/emulators.dart (hunk 3 above)
Comment (devoncarew) [reply packages/flutter_tools/lib/src/commands/emulators.dart]: > If we pick the name for the user then it feels weird if we tell them "an emulator with that name already exists"

We could choose a name, see if it already exists, and suffix something to de-dup (my_name-2).

> and then let them pick a device

I generally like to avoid asking the user for input on stdin. I think `flutter emulators --create --name=my_device` would be a reasonable UI, but am not fixed on it if you think there's a better way.

File: packages/flutter_tools/lib/src/android/android_sdk.dart (hunk 2 above)
Comment (DanTup) [reply packages/flutter_tools/lib/src/android/android_sdk.dart]: Ok, I'll remove the `which`. Looks like there are still some code paths that don't do what I described above, including `getAdbDevices`, so my comment about was a little misleading. Whether or not `_checkForSupportedAdbVersion` is behaving as we'd like though, I'm not sure. LMK if you think this needs some investigation and I'll take a look outside of this PR.

File: packages/flutter_tools/lib/src/commands/emulators.dart (hunk 3 above)
Comment (DanTup) [reply packages/flutter_tools/lib/src/commands/emulators.dart]: Ok, gone with this. We only add suffixes if we're generating the name, so if the user provides an explicit name we'll just give them an error:

```
dantup-macbookpro:flutter dantup$ flutter emulator --create
Emulator 'flutter_emulator' created successfully.

dantup-macbookpro:flutter dantup$ flutter emulator --create
Emulator 'flutter_emulator_1' created successfully.

dantup-macbookpro:flutter dantup$ flutter emulator --create --name flutter_emulator_1
Failed to create emulator 'flutter_emulator_1'.
Error: Android Virtual Device 'flutter_emulator_1' already exists.
```

It feels a bit weird doing `flutter emulator --create` and not `flutter emulator create` but I think that's unavoidable as-is.

Do we have any convention for indicating when a flag is only usable in conjunction with another (eg. `--name` only makes sense with `--create`)? I couldn't see anything obvious.

Comment (devoncarew) [reply packages/flutter_tools/lib/src/commands/emulators.dart]: > Do we have any convention for indicating when a flag is only usable in conjunction with another (eg. --name only makes sense with --create)? I couldn't see anything obvious

Not that I know of; likely just something we should do w/ the help text for the flags.

File: packages/flutter_tools/lib/src/commands/emulators.dart (hunk 4 above)
Comment (devoncarew) [reply packages/flutter_tools/lib/src/commands/emulators.dart]: This looks great for error text. We'll probably also want something brief for `flutter emulators -h` help text as well - perhaps pointing to a flutter.io page?

@mit-mit, @mjohnsullivan - we may want to consider a brief landing page on flutter.io describing how to create android emulators, or at least just routing people to places to find for info about it.

**Comment by devoncarew** [comment]
> This still badly needs some tests - assuming the general idea is ok I'll work on them this week.

Idea looks great!

File: packages/flutter_tools/lib/src/commands/emulators.dart (hunk 3 above)
Comment (DanTup) [reply packages/flutter_tools/lib/src/commands/emulators.dart]: I found some are prefixed with `Used with flag --xxx.` so I'll copy that.

**Comment by DanTup** [comment]
Ok, I've pushed some tests for this. It turned out to be a bit more involved than I expected - I had to make some changes to the existing code to use processManager (so it can be mocked) plus some other tweaks. I tried to split the commits up to make it easier to understand all the changes.

I don't know if this is all the best way to do this (I've not used Mockito before and I'm not sure of our preferences between mocked and real integration tests) but this basically mocks the process manager to return canned responses instead of calling real SDK commands (I wasn't sure how well that's work in the build environment). Please review carefully and nitpick away!

I'm going to do some additional manual testing on Windows shortly, but seemed worthwhile kicking the CI run off to get the results quicker.

**Comment by DanTup** [comment]
Apparently this code had many issues! :(

I've fixed a bunch of bugs from testing on Windows and on the build servers (eg. due to the avd folder didn't exist) and improved some errors. We're touching a lot of stuff so I suspect we may find more edge cases at it runs in more places.

After this, I think it'd might be good to have a real end-to-end test that does something like:

- create emulator
- launch emulator
- check a new device appeared in `flutter devices`
- maybe even launch an app on it?

I'm not sure how fragile that might be (or how well the build infrastructure will support it - I'd guess the devicelab machines should be able to do it?) but it might be a good check.

**Comment by DanTup** [comment]
I seem to get a lot of issues like this on build servers:

```
00:18 +109: test/emulator_test.dart: EmulatorManager create emulator with an empty name does not fail                                                                                                  type 'MappedListIterable' is not a subtype of type 'List<int>' of 'availableApiVersions'
```

Yet get no warnings locally, even with `--preview-dart-2` passed to VM and analyzer. Not sure if I'm doing something wrong?

File: packages/flutter_tools/lib/src/commands/emulators.dart [hunk 5]
```diff
@@ -70,17 +77,27 @@ class EmulatorsCommand extends FlutterCommand {
     }
   }
 
+  Future<Null> _createEmulator(String name) async {
```
Comment (devoncarew) [inline packages/flutter_tools/lib/src/commands/emulators.dart]: Shouldn't `name` be optional here?

File: packages/flutter_tools/lib/src/emulator.dart [hunk 6]
```diff
@@ -57,6 +60,136 @@ class EmulatorManager {
...
+      int suffix = 1;
+      name = autoName;
+      while (takenNames.contains(name)) {
+        name = '${autoName}_${++suffix}';
+      }
+    }
+    
+    final String device =  await _getPreferredAvailableDevice();
+    if (device == null)
+      return new CreateEmulatorResult(
```
Comment (devoncarew) [inline packages/flutter_tools/lib/src/emulator.dart]: nit: the formatting here looks off. 2 space indent? no trailing comma?

Since this is a newer file - added by you recently - you may just want to dartfmt it as a level-set.

File: packages/flutter_tools/lib/src/emulator.dart [hunk 7]
```diff
@@ -57,6 +60,136 @@ class EmulatorManager {
...
+        name,
+        success: false,
+        error: 'No device definitions are available'
+    );
+
+    final String sdkId =  await _getPreferredSdkId();
```
Comment (devoncarew) [inline packages/flutter_tools/lib/src/emulator.dart]: nit: two spaces before await.

File: packages/flutter_tools/lib/src/commands/emulators.dart [hunk 8]
```diff
@@ -16,13 +16,18 @@ class EmulatorsCommand extends FlutterCommand {
   EmulatorsCommand() {
     argParser.addOption('launch',
         help: 'The full or partial ID of the emulator to launch.');
+    argParser.addFlag('create',
+        help: 'Creates a new Android emulator.',
```
Comment (devoncarew) [inline packages/flutter_tools/lib/src/commands/emulators.dart]: We may want to make this a bit longer; we could say that we'll try and create a pixel device, and perhaps include a goo.gl shortened link for more info (until we have a short link to a flutter.io page).

File: packages/flutter_tools/lib/src/emulator.dart (hunk 6 above)
Comment (DanTup) [reply packages/flutter_tools/lib/src/emulator.dart]: Done (though I tweaked some of its changes where they seemed to make it less readable).

File: packages/flutter_tools/lib/src/commands/emulators.dart (hunk 8 above)
Comment (DanTup) [reply packages/flutter_tools/lib/src/commands/emulators.dart]: I added "based on a Pixel device" though not sure what more to add. Do we have somewhere reasonable to link to? The current links we have (in the help) go to pages on managing AVDs with the Android SDK which might not be totally appropriate here?

File: packages/flutter_tools/lib/src/commands/emulators.dart (hunk 5 above)
Comment (devoncarew) [reply packages/flutter_tools/lib/src/commands/emulators.dart]: Cool, I think it just makes it a bit more explicit that it can be null (in the absence of non-null annotations).

File: packages/flutter_tools/lib/src/commands/emulators.dart [hunk 9]
```diff
@@ -77,7 +77,7 @@ class EmulatorsCommand extends FlutterCommand {
     }
   }
 
-  Future<Null> _createEmulator(String name) async {
+  Future<Null> _createEmulator([String name]) async {
```
Comment (devoncarew) [inline packages/flutter_tools/lib/src/commands/emulators.dart]: As a nit, I generally use named params.

**Review by devoncarew — APPROVED** [review]
